File: apps/web/src/types/paper_trading.ts
export interface PaperTradingSummary {
  initial_capital: number;
  current_capital: number;
  invested_value: number;
  unrealized_pnl: number;
  realized_pnl: number;
  total_pnl: number;
  total_pnl_pct: number;
  open_positions: number;
  total_trades: number;
  winning_trades: number;
  win_rate: number;
}

export interface PaperTradingPerformance extends PaperTradingSummary {
  losing_trades: number;
  avg_win_pct: number;
  avg_loss_pct: number;
  best_trade_pct: number;
  worst_trade_pct: number;
}

export interface Position {
  id: string;
  symbol: string;
  side: "BUY";
  quantity: number;
  avg_entry: number;
  stop_loss: number | null;
  t1: number | null;
  t2: number | null;
  t3: number | null;
  trailing_sl: number | null;
  current_sl_level: string;
  signal_id: string | null;
  opened_at: string;
  status: "open" | "partially_closed" | "closed";
  notes: string | null;
  auto_created: boolean;
  creation_source: "manual" | "scanner_auto";
  // Live-enriched fields
  current_price?: number | null;
  unrealized_pnl?: number | null;
  unrealized_pnl_pct?: number | null;
}

export interface Trade {
  id: string;
  position_id: string | null;
  symbol: string;
  side: "BUY" | "SELL";
  quantity: number;
  price: number;
  executed_at: string;
  exit_reason: string | null;
  pnl_abs: number | null;
  pnl_pct: number | null;
  notes: string | null;
}

File: backend/app/models/paper_trading.py
from pydantic import BaseModel, UUID4, field_validator
from datetime import datetime
from typing import Optional


class SellRequest(BaseModel):
    position_id: str
    quantity: int
    price: float
    exit_reason: str = "manual"
    notes: Optional[str] = None

    @field_validator("exit_reason")
    @classmethod
    def validate_reason(cls, v: str) -> str:
        allowed = {"manual", "sl_hit", "t1_hit", "t2_hit", "t3_hit", "trail"}
        if v not in allowed:
            raise ValueError(f"exit_reason must be one of {allowed}")
        return v


class PositionModel(BaseModel):
    id: UUID4
    symbol: str
    side: str
    quantity: int
    avg_entry: float
    stop_loss: Optional[float] = None
    t1: Optional[float] = None
    t2: Optional[float] = None
    t3: Optional[float] = None
    trailing_sl: Optional[float] = None
    current_sl_level: str
    signal_id: Optional[UUID4] = None
    opened_at: datetime
    status: str
    notes: Optional[str] = None
    auto_created: bool = False
    creation_source: str = "manual"
    # Computed at query time
    current_price: Optional[float] = None
    unrealized_pnl: Optional[float] = None
    unrealized_pnl_pct: Optional[float] = None


class TradeModel(BaseModel):
    id: UUID4
    position_id: Optional[UUID4] = None
    symbol: str
    side: str
    quantity: int
    price: float
    executed_at: datetime
    exit_reason: Optional[str] = None
    pnl_abs: Optional[float] = None
    pnl_pct: Optional[float] = None
    notes: Optional[str] = None


class PaperTradingSummary(BaseModel):
    initial_capital: float
    current_capital: float
    invested_value: float
    unrealized_pnl: float
    realized_pnl: float
    total_pnl: float
    total_pnl_pct: float
    open_positions: int
    total_trades: int
    winning_trades: int
    win_rate: float
